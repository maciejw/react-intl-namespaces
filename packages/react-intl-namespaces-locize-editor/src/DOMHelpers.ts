function delay(timeout: number = 0) {
  return new Promise<void>(resolve => {
    window.setTimeout(resolve, timeout);
  });
}

export namespace DOMHelpers {
  export namespace Editor {
    export function toggleClass(
      element: HTMLElement,
      state: boolean,
      cssClass: string,
    ) {
      if (state) {
        element.classList.add(cssClass);
      } else {
        element.classList.remove(cssClass);
      }
    }
    export function createTargetElement(document: Document, id: string) {
      let editor = document.getElementById(id);
      if (editor == null) {
        editor = document.createElement('div');
        editor.id = id;
        document.body.appendChild(editor);
      }
      return editor;
    }
    export function isHtmlElement(element: EventTarget): element is HTMLElement;
    export function isHtmlElement(element: any): element is HTMLElement {
      return element instanceof HTMLElement;
    }

    export function findElementInDOMHierarchy(
      element: Element,
      filter: (htmlElement: HTMLElement) => boolean,
    ) {
      if (isHtmlElement(element)) {
        let e = element;
        while (e !== e.ownerDocument.body) {
          if (filter(e)) {
            return true;
          }
          if (e.parentElement === null) {
            return false;
          }
          e = e.parentElement;
        }
      }
      return false;
    }
    export function ignoreEditorOrTranslatedFilter(e: HTMLElement) {
      return (
        e.dataset.ignoreEditor === 'true' || e.dataset.translated === 'true'
      );
    }
    function isTranslatedOrIgnored(element: Element) {
      return findElementInDOMHierarchy(
        element,
        DOMHelpers.Editor.ignoreEditorOrTranslatedFilter,
      );
    }

    export function locateResourceContext(element: EventTarget) {
      if (isHtmlElement(element)) {
        const resourceContainer = element.parentElement;

        if (!resourceContainer || isTranslatedOrIgnored(resourceContainer)) {
          return false;
        }

        const {
          namespace,
          key,
          defaultMessage,
          description,
        } = resourceContainer.dataset;

        if (namespace && key) {
          return { namespace, key, defaultMessage, description };
        }
      }
      return false;
    }
  }

  export namespace EditorWindow {
    export function postMessage(
      window: Window | null,
      message: any,
      targetOrigin: string,
      transfer?: any[],
    ) {
      if (window !== null) {
        window.postMessage(message, targetOrigin, transfer);
      }
    }

    export function focus(window: Window | null) {
      if (window !== null) {
        window.focus();
      }
    }

    export async function openIfNecessary(
      window: Window,
      openedWindow: Window | null,
      windowOpenTimeout: number,
      url: string,
      target?: string,
      features?: string,
      replace?: boolean,
    ) {
      if (openedWindow === null || openedWindow.closed) {
        openedWindow = window.open(url, target, features, replace);
        await delay(windowOpenTimeout);
      }

      return openedWindow;
    }
  }
}
