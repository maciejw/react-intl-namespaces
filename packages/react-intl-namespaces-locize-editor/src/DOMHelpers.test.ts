import { DOMHelpers } from './DOMHelpers';

describe('DomHelpers', () => {
  const WindowMock = jest.fn<Window>(({ closed = false } = {}) => ({
    closed,
    focus: jest.fn(),
    open: jest.fn(),
    postMessage: jest.fn(),
  }));

  it('should focus', () => {
    const windowMock = new WindowMock();

    DOMHelpers.EditorWindow.focus(windowMock);

    expect(windowMock).toMatchSnapshot();
  });

  it('should post message', () => {
    const windowMock = new WindowMock();

    DOMHelpers.EditorWindow.postMessage(
      windowMock,
      { message: 'value' },
      'some-target',
      ['transfer', 'params'],
    );

    expect(windowMock).toMatchSnapshot();
  });
  it('should open window after delay if its closed', async () => {
    const windowMock = new WindowMock({ closed: true });

    const openedWindow = await DOMHelpers.EditorWindow.openIfNecessary(
      windowMock,
      windowMock,
      0,
      'some-url',
      'some-target',
      'some-feature',
      true,
    );

    expect(windowMock).toMatchSnapshot();
  });
  it('should not locate resource context when target is not html element', () => {
    const result = DOMHelpers.Editor.locateResourceContext(document);
    expect(result).toBe(false);
  });

  it('should not find in dom hierarchy when parent element is null', () => {
    const divWithoutParent = document.createElement('div');
    const result = DOMHelpers.Editor.findElementInDOMHierarchy(
      divWithoutParent,
      e => false,
    );
    expect(result).toBe(false);
  });
  it('should quit searching in dom hierarchy when body is reached', () => {
    const divWithoutParent = document.createElement('div');

    const filterMock = jest.fn();
    const result = DOMHelpers.Editor.findElementInDOMHierarchy(
      document.body,
      filterMock,
    );
    expect(result).toBe(false);
    expect(filterMock).not.toBeCalled();
  });
});
