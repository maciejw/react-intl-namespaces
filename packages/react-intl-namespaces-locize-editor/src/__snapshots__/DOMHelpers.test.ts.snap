// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomHelpers should focus 1`] = `
Object {
  "closed": false,
  "focus": [MockFunction] {
    "calls": Array [
      Array [],
    ],
  },
  "open": [MockFunction],
  "postMessage": [MockFunction],
}
`;

exports[`DomHelpers should open window after delay if its closed 1`] = `
Object {
  "closed": true,
  "focus": [MockFunction],
  "open": [MockFunction] {
    "calls": Array [
      Array [
        "some-url",
        "some-target",
        "some-feature",
        true,
      ],
    ],
  },
  "postMessage": [MockFunction],
}
`;

exports[`DomHelpers should post message 1`] = `
Object {
  "closed": false,
  "focus": [MockFunction],
  "open": [MockFunction],
  "postMessage": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "message": "value",
        },
        "some-target",
        Array [
          "transfer",
          "params",
        ],
      ],
    ],
  },
}
`;
